### PR TITLE
[pretyper] Remove imperative passing of evar_map.

### PR DESCRIPTION
**Pull request by ejgallego**

### CODE REVIEW

**Comment by ejgallego** [comment]
If consider it opportune you can merge this at once and close #8580

**Comment by ejgallego** [comment]
Bench here https://ci.inria.fr/coq/job/benchmark-part-of-the-branch/525/console

Maybe flambda could do better? Anyways I guess this codepath is not so critical in the bench.

**Comment by ejgallego** [comment]
```
┌──────────────────────────┬─────────────────────────┬───────────────────────────────────────┬───────────────────────────────────────┬─────────────────────────┬─────────────────────┐
│                          │      user time [s]      │              CPU cycles               │           CPU instructions            │  max resident mem [KB]  │     mem faults      │
│                          │                         │                                       │                                       │                         │                     │
│             package_name │     NEW     OLD PDIFF   │            NEW            OLD PDIFF   │            NEW            OLD PDIFF   │     NEW     OLD PDIFF   │  NEW  OLD   PDIFF   │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────────┤
│      coq-formal-topology │   54.48   54.90 -0.77 % │   148616520326   148943673583 -0.22 % │   226377919991   226485860243 -0.05 % │  479996  479972 +0.01 % │    0    0    +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────────┤
│                coq-sf-lf │   40.66   40.91 -0.61 % │   112159045289   112438959472 -0.25 % │   185849415773   185914678331 -0.04 % │  422952  422904 +0.01 % │  128    0    +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────────┤
│   coq-mathcomp-odd-order │ 1410.17 1416.18 -0.42 % │  3924333700459  3940929074251 -0.42 % │  6598128333865  6597510074053 +0.01 % │ 1356936 1361032 -0.30 % │    2    0    +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────────┤
│                 coq-hott │  329.28  329.88 -0.18 % │   917765432289   919207409776 -0.16 % │  1381043070738  1381369199135 -0.02 % │  610240  610312 -0.01 % │  247    0    +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────────┤
│           coq-coquelicot │   99.94  100.05 -0.11 % │   276825473721   276796694768 +0.01 % │   373126041747   373020143056 +0.03 % │  716120  715888 +0.03 % │  281   14 +1907.14 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────────┤
│     coq-mathcomp-algebra │  201.70  201.87 -0.08 % │   559904025600   559935311004 -0.01 % │   787139379573   787411216450 -0.03 % │  645652  647612 -0.30 % │   21    6 +250.00 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────────┤
│       coq-mathcomp-field │  464.13  464.51 -0.08 % │  1290981987097  1292425065734 -0.11 % │  2113261715891  2113242126792 +0.00 % │  805692  805944 -0.03 % │    0   24 -100.00 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────────┤
│ coq-mathcomp-real-closed │  184.72  184.87 -0.08 % │   514169505586   514462924783 -0.06 % │   779884714671   779915409839 -0.00 % │  829992  830168 -0.02 % │    0    0    +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────────┤
│                  coq-vst │ 1688.55 1689.62 -0.06 % │  4694622731374  4696151591694 -0.03 % │  6428615323505  6429505813900 -0.01 % │ 2226592 2226472 +0.01 % │  392  271  +44.65 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────────┤
│               coq-sf-plf │   70.81   70.85 -0.06 % │   195640948307   196058308247 -0.21 % │   286775993317   286897662398 -0.04 % │  516876  516864 +0.00 % │   15    0    +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────────┤
│   coq-mathcomp-ssreflect │   62.10   62.12 -0.03 % │   170616986181   170693174744 -0.04 % │   243909388732   243979802393 -0.03 % │  535096  534972 +0.02 % │    2    0    +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────────┤
│         coq-math-classes │  243.47  243.33 +0.06 % │   671338712846   670026988054 +0.20 % │   877812410022   877913265757 -0.01 % │  524736  524596 +0.03 % │   84   84   +0.00 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────────┤
│    coq-mathcomp-solvable │  216.99  216.79 +0.09 % │   602010820113   601661023910 +0.06 % │   860387625664   860455791291 -0.01 % │  865736  838652 +3.23 % │    0  203 -100.00 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────────┤
│              coq-bignums │   95.23   95.14 +0.09 % │   263692918800   262801392742 +0.34 % │   376545906591   376701960664 -0.04 % │  540392  540308 +0.02 % │   16    0    +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────────┤
│                coq-color │  732.73  731.79 +0.13 % │  2029347682879  2026088482913 +0.16 % │  2440247694082  2439820942792 +0.02 % │ 1395136 1389872 +0.38 % │   39   29  +34.48 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────────┤
│    coq-mathcomp-fingroup │   81.58   81.43 +0.18 % │   225038319852   224972671870 +0.03 % │   336130301011   336063668044 +0.02 % │  592348  590156 +0.37 % │  154    0    +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────────┤
│              coq-unimath │ 1323.31 1320.73 +0.20 % │  3677288693420  3668036192213 +0.25 % │  6216520700051  6212621538229 +0.06 % │ 1053576 1053176 +0.04 % │  248   20 +1140.00 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────────┤
│            coq-fiat-core │  131.62  131.36 +0.20 % │   366084401083   365920319778 +0.04 % │   488905429950   488693855042 +0.04 % │  501840  501916 -0.02 % │   33   68  -51.47 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────────┤
│                coq-gpfsl │ 1192.51 1189.88 +0.22 % │  3322041972620  3315120226981 +0.21 % │  4372131854828  4367339812241 +0.11 % │ 1602704 1585632 +1.08 % │   58    0    +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────────┤
│   coq-mathcomp-character │  289.79  289.02 +0.27 % │   805882723973   803656827606 +0.28 % │  1172574031748  1172596334515 -0.00 % │ 1118648 1118476 +0.02 % │    0    0    +nan % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────────┤
│         coq-fiat-parsers │  714.73  712.71 +0.28 % │  1983716667984  1977599523738 +0.31 % │  3036873713043  3036252900487 +0.02 % │ 2791036 2793228 -0.08 % │   25  316  -92.09 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────────┤
│                 coq-corn │ 1548.72 1542.54 +0.40 % │  4294380475652  4280217881880 +0.33 % │  6336248685642  6336440993997 -0.00 % │  858420  858648 -0.03 % │   44   44   +0.00 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────────┤
│               coq-geocoq │ 2826.68 2812.65 +0.50 % │  7858801839463  7819062277133 +0.51 % │ 12527482063774 12526915209118 +0.00 % │ 1390572 1391048 -0.03 % │   92   82  +12.20 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────────┤
│          coq-fiat-crypto │ 6316.68 6283.28 +0.53 % │ 17559646299048 17479654545865 +0.46 % │ 28517646672952 28499127084267 +0.06 % │ 2533524 2589376 -2.16 % │ 1399 1152  +21.44 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────────┤
│                coq-flocq │  585.16  581.96 +0.55 % │  1628526224200  1619930045016 +0.53 % │  2356912349384  2357521800249 -0.03 % │ 1712808 1712404 +0.02 % │  207   46 +350.00 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────────┤
│             coq-compcert │  888.84  882.47 +0.72 % │  2464411743125  2445496297547 +0.77 % │  3539687936991  3540464552435 -0.02 % │ 1341320 1342176 -0.06 % │  103  156  -33.97 % │
├──────────────────────────┼─────────────────────────┼───────────────────────────────────────┼───────────────────────────────────────┼─────────────────────────┼─────────────────────┤
│               coq-sf-vfa │   44.28   43.89 +0.89 % │   121452181638   121302857606 +0.12 % │   196522511015   196593983299 -0.04 % │  537488  537472 +0.00 % │   28    5 +460.00 % │
└──────────────────────────┴─────────────────────────┴───────────────────────────────────────┴───────────────────────────────────────┴─────────────────────────┴─────────────────────┘
```

**Review by ppedrot — APPROVED** [review]
LGTM. Bench is within noise bounds.

**Comment by ejgallego** [comment]
An interesting experiment I am trying is to compare how much flambda can improve using this PR vs not having it. It is too early to say but I'd dare to say that differences are not noise:

- Recent Coq, flambda vs not-flambda, same commit: https://ci.inria.fr/coq/job/benchmark-part-of-the-branch/513/console
- Coq before this PR, non-flambda vs this PR + flabmda: https://ci.inria.fr/coq/job/benchmark-part-of-the-branch/529/console [ongoing]

It is a pity we cannot use 4.07.0 due to our dependency on batteries.

[ccing @gasche as he is usually interested in flambda numbers].

**Comment by SkySkimmer** [comment]
How do we depend on batteries and how does it stop 07?

**Comment by ejgallego** [comment]
> How do we depend on batteries and how does it stop 07?

Printing of the bench tables depends on batteries, and for your second question https://opam.ocaml.org/packages/batteries/

**Comment by ppedrot** [comment]
> It is too early to say but I'd dare to say that differences are not noise

Does that mean that with flambda, in general, writing pure code is more efficient than the corresponding impure one? That'd be a good thing.

**Comment by ejgallego** [comment]
> My understanding is that indeed the optimizer can be far more aggressive when functional code is present; references place significant barriers to it I understand.

That's a reason indeed to try to get rid for example of all the `Store.t` we currently have in `evar_map` / `evar_info`.

**Comment by ejgallego** [comment]
Bench done:
```
┌──────────────────────────┬──────────────────────────┬──────────────────────────┬
│                          │      user time [s]       │      user time [s]       │
│                          │                          │                          │
│             package_name │     NEW     OLD  PDIFF   │     NEW     OLD  PDIFF   │
├──────────────────────────┼──────────────────────────┼──────────────────────────┼
│               coq-geocoq │ 2472.38 2782.58 -11.15 % │ 2477.66 2832.50 -12.53 % │
├──────────────────────────┼──────────────────────────┼──────────────────────────┼
│          coq-lambda-rust │ 1928.30 2158.46 -10.66 % │ FAILED to INSTALL        │
├──────────────────────────┼──────────────────────────┼──────────────────────────┼
│                  coq-vst │ 1464.25 1634.21 -10.40 % │ 1488.25 1695.97 -12.25 % │
├──────────────────────────┼──────────────────────────┼──────────────────────────┼
│          coq-fiat-crypto │ 5568.47 6212.45 -10.37 % │ 5582.27 6310.92 -11.55 % │
├──────────────────────────┼──────────────────────────┼──────────────────────────┼
│              coq-unimath │ 1189.04 1319.50  -9.89 % │ 1190.66 1322.86  -9.99 % │
├──────────────────────────┼──────────────────────────┼──────────────────────────┼
│                 coq-corn │ 1371.22 1510.48  -9.22 % │ 1376.67 1551.34 -11.26 % │
├──────────────────────────┼──────────────────────────┼──────────────────────────┼
│             coq-compcert │  787.45  865.58  -9.03 % │  785.44  886.01 -11.35 % │
├──────────────────────────┼──────────────────────────┼──────────────────────────┼
│                coq-color │  640.83  703.49  -8.91 % │  640.66  733.19 -12.62 % │
├──────────────────────────┼──────────────────────────┼──────────────────────────┼
│   coq-mathcomp-odd-order │ 1278.50 1402.62  -8.85 % │ 1285.86 1413.12  -9.01 % │
├──────────────────────────┼──────────────────────────┼──────────────────────────┼
│   coq-mathcomp-character │  262.85  288.17  -8.79 % │  264.70  290.97  -9.03 % │
├──────────────────────────┼──────────────────────────┼──────────────────────────┼
│       coq-mathcomp-field │  423.11  462.22  -8.46 % │  423.58  465.83  -9.07 % │
├──────────────────────────┼──────────────────────────┼──────────────────────────┼
│ coq-mathcomp-real-closed │  169.22  184.08  -8.07 % │  169.09  185.37  -8.78 % │
├──────────────────────────┼──────────────────────────┼──────────────────────────┼
│                 coq-hott │  301.52  327.98  -8.07 % │  309.81  338.89  -8.58 % │
├──────────────────────────┼──────────────────────────┼──────────────────────────┼
│              coq-bignums │   86.89   94.35  -7.91 % │   86.38   95.32  -9.38 % │
├──────────────────────────┼──────────────────────────┼──────────────────────────┼
│    coq-mathcomp-solvable │  197.89  214.63  -7.80 % │  199.04  217.42  -8.45 % │
├──────────────────────────┼──────────────────────────┼──────────────────────────┼
│         coq-fiat-parsers │  647.44  699.42  -7.43 % │  642.03  713.42 -10.01 % │
├──────────────────────────┼──────────────────────────┼──────────────────────────┼
│         coq-math-classes │  218.50  235.78  -7.33 % │  221.46  245.32  -9.73 % │
├──────────────────────────┼──────────────────────────┼──────────────────────────┼
│     coq-mathcomp-algebra │  185.56  200.18  -7.30 % │  185.60  202.70  -8.44 % │
├──────────────────────────┼──────────────────────────┼──────────────────────────┼
│                coq-flocq │  534.40  566.42  -5.65 % │  546.51  582.99  -6.26 % │
├──────────────────────────┼──────────────────────────┼──────────────────────────┼
│    coq-mathcomp-fingroup │   76.46   80.69  -5.24 % │   76.66   81.66  -6.12 % │
├──────────────────────────┼──────────────────────────┼──────────────────────────┼
│               coq-sf-plf │   65.73   69.19  -5.00 % │   65.70   71.34  -7.91 % │
├──────────────────────────┼──────────────────────────┼──────────────────────────┼
│           coq-coquelicot │   92.26   96.92  -4.81 % │   92.03  100.29  -8.24 % │
├──────────────────────────┼──────────────────────────┼──────────────────────────┼
│            coq-fiat-core │  120.28  123.54  -2.64 % │  119.73  132.08  -9.35 % │
├──────────────────────────┼──────────────────────────┼──────────────────────────┼
│      coq-formal-topology │   52.41   53.74  -2.47 % │   52.64   54.82  -3.98 % │
├──────────────────────────┼──────────────────────────┼──────────────────────────┼
│   coq-mathcomp-ssreflect │   59.62   61.11  -2.44 % │   59.98   62.30  -3.72 % │
├──────────────────────────┼──────────────────────────┼──────────────────────────┼
│               coq-sf-vfa │   42.10   42.62  -1.22 % │   42.30   44.18  -4.26 % │
├──────────────────────────┼──────────────────────────┼──────────────────────────┼
│                coq-sf-lf │   39.62   39.29  +0.84 % │   39.71   41.10  -3.38 % │
└──────────────────────────┴──────────────────────────┴──────────────────────────┼
```

**Comment by SkySkimmer** [comment]
How do I read this table?

Gaëtan Gilbert

**Comment by ejgallego** [comment]
> How do I read this table?

As indicated above:
- first column: improvement of flambda vs non-flambda, same commit.
- second column: improvement of flambda + this PR vs non-flambda master.

**Comment by ejgallego** [comment]
By the way coq master seems to have gotten slower in the last weeks, as you can witness in the absolute OLD times.

**Comment by ejgallego** [comment]
No point in keeping two PRs open for this; now all that we need is an assignee.

**Comment by ppedrot** [comment]
I will merge soon in the absence of complains.

**Comment by gasche** [comment]
(There is a 4.07.0-compatible release of Batteries in the pipeline, but it has been delayed by the opam2 repo migration. @ejgallego, in the meantime feel free to pin https://github.com/ocaml-batteries-team/batteries-included/releases/tag/v2.9.0 directly.

File: clib/cArray.ml
Comment (herbelin) [inline clib/cArray.ml]: Is this more efficient with a ref?

**Review by herbelin — APPROVED** [review]
I have the same patch to `globEnv` as part of #7300, but that's not a problem.

File: clib/cArray.mli
Comment (gasche) [inline clib/cArray.mli]: Did you mean `Same as [fold_left2_map]`?

Comment (ejgallego) [reply clib/cArray.mli]: Ups, fixed thanks!

File: clib/cArray.ml
Comment (ejgallego) [reply clib/cArray.ml]: I just copied the code from the non `_i` version. I guess that's a question for experts in low-level OCaml code.

Comment (gasche) [reply clib/cArray.ml]: Well, given that you just observed non-trivial flambda speedups by removing writes to references, it makes sense to wonder whether avoiding them in those library functions (which sounds quite easy) would again improve performances.

(In general writes to reference can slow things down when they hit the write barrier, unless their static type guarantees that the references don't contain pointers (`int ref` for example). I don't remember offhand if short-lived references are also affected, or only long-lived references. In the context of this PR, it seems that what you observe is rather than flambda is better at optimizing tuple-passing than reference-read-writing code. I think this second effect would not be observed here, unless both `fold_left2_map_i` and `map2_i` were inlined at the point of use -- which you may be able to tell by looking at flambda inliner's reports, but I cannot guess from my chair.)

Comment (ejgallego) [reply clib/cArray.ml]: > Well, given that you just observed non-trivial flambda speedups by removing writes to references, it makes sense to wonder whether avoiding them in those library functions (which sounds quite easy) would again improve performances.

IIRC the main reason to use imperative style in these functions may be due to lack of tail-call, but again I am not an expert, so I'll let others handle this.

**Comment by ppedrot** [comment]
Meh, this now needs a rebase before I merge.

**Comment by ejgallego** [comment]
> Meh, this now needs a rebase before I merge.

Rebase was not trivial so I suggest we let the CI run.

**Comment by ejgallego** [comment]
> Rebase was not trivial so I suggest we let the CI run.

CI is basically green so we are good here.

**Comment by ejgallego** [comment]
Merci !

**Pull request merged**
